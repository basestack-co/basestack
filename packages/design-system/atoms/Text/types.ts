import { ReactNode } from "react";
import { SpaceProps } from "styled-system";

export type Size =
  | "xxLarge"
  | "xLarge"
  | "large"
  | "medium"
  | "small"
  | "xSmall";

export interface TextProps extends SpaceProps {
  /**
   * Change text size
   */
  size?: Size;
  /**
   * Changes text to gray
   */
  muted?: boolean;
  /**
   * Change text color
   */
  color?: string;
  /**
   * Text
   */
  children: ReactNode;
}
