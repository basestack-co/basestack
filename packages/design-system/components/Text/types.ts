import { ReactNode } from "react";
import { SpaceProps, TypographyProps, FlexboxProps } from "styled-system";

export type Size =
  | "xxLarge"
  | "xLarge"
  | "large"
  | "medium"
  | "small"
  | "xSmall";

export type FontFamily = "roboto" | "robotoFlex";

export interface TextProps extends SpaceProps, TypographyProps, FlexboxProps {
  /**
   * Change text tag
   */
  as?: string;
  /**
   * Change text size
   */
  size?: Size;
  /**
   * Changes text to gray
   */
  muted?: boolean;
  /**
   * Change text color
   */
  color?: string;
  /**
   * Text
   */
  children: ReactNode;
  /**
   * Change text fontFamily
   */
  fontFamily?: FontFamily;
  /**
   * Truncate text to one line
   */
  lineTruncate?: boolean;
}
