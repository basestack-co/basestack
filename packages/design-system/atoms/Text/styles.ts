import styled, { css } from "styled-components";
import { space } from "styled-system";
import { rem } from "polished";
import { TextProps } from "./types";

const sharedStyles = (color?: string, muted = false) => css`
  ${space};
  font-family: ${({ theme }) => theme.typography.fontFamily};
  color: ${({ theme }) =>
    color || (muted ? theme.colors.gray600 : theme.colors.black)};
`;

export const XSmallText = styled.span<TextProps>`
  display: flex;
  ${({ color, muted }) => sharedStyles(color, muted)};
  font-size: ${rem("12px")};
  font-weight: 400;
`;

export const SmallText = styled.p<TextProps>`
  ${({ color, muted }) => sharedStyles(color, muted)};
  font-size: ${rem("14px")};
  font-weight: 400;
`;

export const MediumText = styled.p<TextProps>`
  ${({ color, muted }) => sharedStyles(color, muted)};
  font-size: ${rem("16px")};
  font-weight: 500;
`;

export const LargeText = styled.h3<TextProps>`
  ${({ color, muted }) => sharedStyles(color, muted)};
  font-size: ${rem("18px")};
  font-weight: 500;
`;

export const XLargeText = styled.h2<TextProps>`
  ${({ color, muted }) => sharedStyles(color, muted)};
  font-size: ${rem("20px")};
  font-weight: 700;
`;

export const XXLargeText = styled.h1<TextProps>`
  ${({ color, muted }) => sharedStyles(color, muted)};
  font-size: ${rem("24px")};
  font-weight: 700;
`;
