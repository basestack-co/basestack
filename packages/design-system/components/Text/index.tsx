import { memo } from "react";
import { TextProps } from "./types";
import {
  XSmallText,
  SmallText,
  MediumText,
  LargeText,
  XLargeText,
  XXLargeText,
} from "./styles";

const Text = ({
  size = "small",
  muted = false,
  color,
  children,
  fontFamily = "roboto",
  lineTruncate = false,
  ...props
}: TextProps) => {
  const sharedProps = {
    muted,
    color,
    fontFamily,
    lineTruncate,
  };

  switch (size) {
    case "xxLarge":
      return (
        <XXLargeText {...sharedProps} {...props}>
          {children}
        </XXLargeText>
      );
    case "xLarge":
      return (
        <XLargeText {...sharedProps} {...props}>
          {children}
        </XLargeText>
      );
    case "large":
      return (
        <LargeText {...sharedProps} {...props}>
          {children}
        </LargeText>
      );
    case "medium":
      return (
        <MediumText {...sharedProps} {...props}>
          {children}
        </MediumText>
      );
    default:
    case "small":
      return (
        <SmallText {...sharedProps} {...props}>
          {children}
        </SmallText>
      );
    case "xSmall":
      return (
        <XSmallText {...sharedProps} {...props}>
          {children}
        </XSmallText>
      );
  }
};

export { type TextProps };

export default memo(Text);
