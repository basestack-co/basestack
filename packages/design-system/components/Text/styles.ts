import { rem } from "polished";
import styled, { css } from "styled-components";
import { compose, flexbox, layout, space, typography } from "styled-system";
import type { FontFamily, TextProps } from "./types";

interface sharedStylesProps {
  color?: string;
  muted?: boolean;
  fontFamily?: FontFamily;
  lineTruncate?: boolean;
}

const shouldForwardProps = [
  "textAlign",
  "lineTruncate",
  "lineHeight",
  "flexShrink",
  "muted",
  "maxWidth",
];

const sharedStyles = ({
  color,
  muted,
  fontFamily,
  lineTruncate,
}: sharedStylesProps) => css`
  font-family: ${({ theme }) =>
    fontFamily === "robotoFlex"
      ? theme.typography.robotoFlex
      : theme.typography.roboto};
  word-break: break-word;
  color: ${({ theme }) =>
    color || (muted ? theme.text.muted : theme.text.color)};
  ${compose(space, typography, flexbox, layout)};

  ${
    lineTruncate &&
    css`
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
  `
  }
`;

export const XSmallText = styled.span.withConfig({
  shouldForwardProp: (prop) => !shouldForwardProps.includes(prop),
})<TextProps>`
  display: flex;
  font-size: ${rem("12px")};
  line-height: ${rem("14px")};
  font-weight: 400;
  ${({ color, muted, fontFamily, lineTruncate }) =>
    sharedStyles({ color, muted, fontFamily, lineTruncate })};
`;

export const SmallText = styled.p.withConfig({
  shouldForwardProp: (prop) => !shouldForwardProps.includes(prop),
})<TextProps>`
  font-size: ${rem("14px")};
  line-height: ${rem("22px")};
  font-weight: 400;
  ${({ color, muted, fontFamily, lineTruncate }) =>
    sharedStyles({ color, muted, fontFamily, lineTruncate })};
`;

export const MediumText = styled.p.withConfig({
  shouldForwardProp: (prop) => !shouldForwardProps.includes(prop),
})<TextProps>`
  font-size: ${rem("16px")};
  line-height: ${rem("24px")};
  font-weight: 500;
  ${({ color, muted, fontFamily, lineTruncate }) =>
    sharedStyles({ color, muted, fontFamily, lineTruncate })};
`;

export const LargeText = styled.h3.withConfig({
  shouldForwardProp: (prop) => !shouldForwardProps.includes(prop),
})<TextProps>`
  font-size: ${rem("18px")};
  line-height: ${rem("26px")};
  font-weight: 500;
  ${({ color, muted, fontFamily, lineTruncate }) =>
    sharedStyles({ color, muted, fontFamily, lineTruncate })};
`;

export const XLargeText = styled.h2.withConfig({
  shouldForwardProp: (prop) => !shouldForwardProps.includes(prop),
})<TextProps>`
  font-size: ${rem("20px")};
  line-height: ${rem("30px")};
  font-weight: 500;
  ${({ color, muted, fontFamily, lineTruncate }) =>
    sharedStyles({ color, muted, fontFamily, lineTruncate })};
`;

export const XXLargeText = styled.h1.withConfig({
  shouldForwardProp: (prop) => !shouldForwardProps.includes(prop),
})<TextProps>`
  font-size: ${rem("24px")};
  line-height: ${rem("36px")};
  font-weight: 700;
  ${({ color, muted, fontFamily, lineTruncate }) =>
    sharedStyles({ color, muted, fontFamily, lineTruncate })};
`;
