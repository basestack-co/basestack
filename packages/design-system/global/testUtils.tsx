import { type RenderOptions, render } from "@testing-library/react";
import type React from "react";
// Styles
import { ThemeProvider } from "styled-components";
import theme from "../theme/lightTheme";

interface Props {
  children: React.ReactNode;
}

const WithThemeProvider = ({ children }: Props) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
);

const renderWithTheme = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "wrapper">,
) => render(ui, { wrapper: WithThemeProvider, ...options });

export { renderWithTheme };
