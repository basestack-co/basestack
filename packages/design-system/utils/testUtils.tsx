import defaultTheme from "@basestack/design-system/theme/lightTheme";
import isPropValid from "@emotion/is-prop-valid";
import { RenderResult, render } from "@testing-library/react";
import React from "react";
import { StyleSheetManager, ThemeProvider } from "styled-components";

const Provider = ({
  children,
  theme = defaultTheme,
}: {
  children: React.ReactNode;
  theme: any;
}) => (
  <ThemeProvider theme={theme}>
    <StyleSheetManager shouldForwardProp={isPropValid}>
      {children}
    </StyleSheetManager>
  </ThemeProvider>
);

export const renderWithTheme = (
  ui: React.ReactElement,
  options?: any,
): RenderResult => render(ui, { wrapper: Provider, ...options });
