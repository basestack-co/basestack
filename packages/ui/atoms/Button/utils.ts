import { Size, Target, Type, Variant } from "@/components/atoms/Button/types";
import { css, DefaultTheme } from "styled-components";
import { darken, rem, transparentize } from "polished";

export const handleButtonSize = (size?: Size) => {
  switch (size) {
    case "large":
      return css`
        height: ${rem("60px")};
        font-size: ${rem("18px")};
        line-height: ${rem("25px")};
        padding: 0 ${rem("40px")};
      `;
    case "medium":
      return css`
        height: ${rem("54px")};
        font-size: ${rem("16px")};
        line-height: ${rem("22px")};
        padding: 0 ${rem("34px")};
      `;
    case "normal":
      return css`
        height: ${rem("48px")};
        font-size: ${rem("16px")};
        line-height: ${rem("22px")};
        padding: 0 ${rem("30px")};
      `;
    default:
    case "small":
      return css`
        height: ${rem("46px")};
        font-size: ${rem("14px")};
        line-height: ${rem("19px")};
        padding: 0 ${rem("24px")};
      `;
    case "xSmall":
      return css`
        height: ${rem("42px")};
        font-size: ${rem("14px")};
        line-height: ${rem("19px")};
        padding: 0 ${rem("20px")};
      `;
  }
};

export const handleButtonsColors = (theme: DefaultTheme) => ({
  primary: {
    outline: {
      color: theme.colors.primary,
      borderColor: theme.colors.primary,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.primary,
    },
  },
  secondary: {
    outline: {
      color: theme.colors.secondary,
      borderColor: theme.colors.secondary,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.secondary,
    },
  },
  tertiary: {
    outline: {
      color: theme.colors.gray3,
      borderColor: theme.colors.gray3,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.primary,
      borderColor: "transparent",
      backgroundColor: theme.colors.gray6,
    },
  },
  neutral: {
    outline: {
      color: theme.colors.gray2,
      borderColor: theme.colors.gray2,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.primary,
      borderColor: "transparent",
      backgroundColor: "transparent",
    },
  },
  success: {
    outline: {
      color: theme.colors.success,
      borderColor: theme.colors.success,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.success,
    },
  },
  warning: {
    outline: {
      color: theme.colors.warning,
      borderColor: theme.colors.warning,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.warning,
    },
  },
  danger: {
    outline: {
      color: theme.colors.error,
      borderColor: theme.colors.error,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.error,
    },
  },
});

export const darkenColors = (
  theme: DefaultTheme,
  amount: number,
  type: Type,
  variant: Variant,
  target: Target
) => {
  return darken(
    amount || 0.1,
    handleButtonsColors(theme)[type || "primary"][variant || "filled"][target]
  );
};

export const lightenColors = (
  theme: DefaultTheme,
  amount: number,
  type: Type,
  variant: Variant,
  target: Target
) => {
  return transparentize(
    amount || 0.6,
    handleButtonsColors(theme)[type || "primary"][variant || "filled"][target]
  );
};
