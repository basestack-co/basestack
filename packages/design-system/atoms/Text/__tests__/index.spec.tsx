import React from "react";
import { cleanup } from "@testing-library/react";
import { rem } from "polished";
import { renderWithTheme } from "../../../utils/testUtils";
import Text from "..";

describe("Text Atom tests", () => {
  afterEach(cleanup);

  test("should render text correctly", () => {
    const { asFragment } = renderWithTheme(<Text>Text</Text>);
    expect(asFragment()).toMatchSnapshot();
  });

  test("should render red text", () => {
    const { getByText } = renderWithTheme(<Text color="red">red</Text>);
    expect(getByText(/red/)).toHaveStyle(`color: red`);
  });

  test("should render muted text", () => {
    const { getByText } = renderWithTheme(<Text muted>muted</Text>);
    expect(getByText(/muted/)).toHaveStyle(`color: #545454`);
  });

  test("should render xSmall text", () => {
    const { getByText } = renderWithTheme(<Text size="xSmall">xSmall</Text>);
    const text = getByText(/xSmall/);
    expect(text).toHaveStyle(`font-size: ${rem("12px")}`);
    expect(text).toHaveStyle(`font-weight: 400`);
    expect(text).toHaveStyle(`color: #000000`);
  });

  test("should render small text", () => {
    const { getByText } = renderWithTheme(<Text size="small">small</Text>);
    const text = getByText(/small/);
    expect(text).toHaveStyle(`font-size: ${rem("14px")}`);
    expect(text).toHaveStyle(`font-weight: 400`);
    expect(text).toHaveStyle(`color: #000000`);
  });

  test("should render medium text", () => {
    const { getByText } = renderWithTheme(<Text size="medium">medium</Text>);
    const text = getByText(/medium/);
    expect(text).toHaveStyle(`font-size: ${rem("16px")}`);
    expect(text).toHaveStyle(`font-weight: 500`);
    expect(text).toHaveStyle(`color: #000000`);
  });

  test("should render large text", () => {
    const { getByText } = renderWithTheme(<Text size="large">large</Text>);
    const text = getByText(/large/);
    expect(text).toHaveStyle(`font-size: ${rem("18px")}`);
    expect(text).toHaveStyle(`font-weight: 500`);
    expect(text).toHaveStyle(`color: #000000`);
  });

  test("should render xLarge text", () => {
    const { getByText } = renderWithTheme(<Text size="xLarge">xLarge</Text>);
    const text = getByText(/xLarge/);
    expect(text).toHaveStyle(`font-size: ${rem("20px")}`);
    expect(text).toHaveStyle(`font-weight: 500`);
    expect(text).toHaveStyle(`color: #000000`);
  });

  test("should render xxLarge text", () => {
    const { getByText } = renderWithTheme(<Text size="xxLarge">xxLarge</Text>);
    const text = getByText(/xxLarge/);
    expect(text).toHaveStyle(`font-size: ${rem("24px")}`);
    expect(text).toHaveStyle(`font-weight: 700`);
    expect(text).toHaveStyle(`color: #000000`);
  });
});
