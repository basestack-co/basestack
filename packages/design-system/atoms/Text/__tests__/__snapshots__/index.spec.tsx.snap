// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Atom tests should render text correctly 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 0.875rem;
  line-height: 1.25rem;
  font-weight: 400;
  font-family: 'Roboto',sans-serif;
  color: #000000;
}

<p
    class="c0"
  >
    Text
  </p>
</DocumentFragment>
`;
